source/base/hdf5.cc

```diff
@@ -0,0 +1,1627 @@
+      std::is_same<Container,
+                   std::vector<typename Container::value_type>>::value,
+      std::vector<hsize_t>>::type
+    get_container_dimensions(const Container &data)
+    {
+      return std::vector<hsize_t>{data.size()};
+    }
+
+
+
+    template <typename Container>
+    typename std::enable_if<
+      std::is_same<Container, Vector<typename Container::value_type>>::value,
+      std::vector<hsize_t>>::type
+    get_container_dimensions(const Container &data)
+    {
+      return std::vector<hsize_t>{data.size()};
+    }
+
+
+
+    template <typename Container>
+    typename std::enable_if<
+      std::is_same<Container,
+                   FullMatrix<typename Container::value_type>>::value,
+      std::vector<hsize_t>>::type
+    get_container_dimensions(const Container &data)
+    {
+      return std::vector<hsize_t>{data.m(), data.n()};
+    }
+
+
+
+    // This function returns the total size of the container.
+    // For a std::vector the function returns int(vector_size)
+    // For a Vector the function returns int(vector_size)
+    // For a FullMatrix the function returns int(rows*columns)
+    template <typename Container>
+    typename std::enable_if<
+      std::is_same<Container,
+                   std::vector<typename Container::value_type>>::value,
+      unsigned int>::type
+    get_container_size(const Container &data)
+    {
+      return static_cast<unsigned int>(data.size());
+    }
+
+
+
+    template <typename Container>
+    typename std::enable_if<
+      std::is_same<Container, Vector<typename Container::value_type>>::value,
+      unsigned int>::type
+    get_container_size(const Container &data)
+    {
+      return static_cast<unsigned int>(data.size());
+    }
+
+
+
+    template <typename Container>
+    typename std::enable_if<
+      std::is_same<Container,
+                   FullMatrix<typename Container::value_type>>::value,
+      unsigned int>::type
+    get_container_size(const Container &data)
+    {
+      return static_cast<unsigned int>(data.m() * data.n());
+    }
+
+
+
+    // This function returns the pointer to the raw data of a container
+    template <typename Container>
+    typename std::enable_if<
+      std::is_same<Container,
+                   std::vector<typename Container::value_type>>::value,
+      void *>::type
+    get_container_pointer(Container &data)
+    {
+      // It is very important to pass the variable "data" by reference otherwise
+      // the pointer will be wrong
+      return data.data();
+    }
+
+
+
+    template <typename Container>
+    typename std::enable_if<
+      std::is_same<Container, Vector<typename Container::value_type>>::value,
+      void *>::type
+    get_container_pointer(Container &data)
+    {
+      // It is very important to pass the variable "data" by reference otherwise
+      // the pointer will be wrong.
+      // Use the first element of FullMatrix to get the pointer to the raw data
+      return &data[0];
```

Comment:
Is this the right way to get the pointer to the raw data of a Vector? I could make this function friend for Vector, but that would pollute the interface of Vector